invoice/src/main/java/org/killbill/billing/invoice/optimizer/InvoiceOptimizerExp.java

```diff
@@ -155,22 +157,45 @@ public boolean apply(final InvoiceItem invoiceItem) {
 
                         // Any cutoff date 't' will return invoices with all items where:
                         // - If IN_ADVANCE, all items where startDate >= t
-                        // - If IN_ARREAR, all items where endDate >= t; however, if item was pro-rated, we recompute the endDt based on the startDt + billingPeriod
+                        // - If IN_ARREAR, all items where endDate >= t; however, in more complex use cases
+                        //  (e.g cancellation in past leading to trailing pro-ration), we want to return the item
+                        // so we compare with an exiting item and check for meaningful changes (end date, plan,...)
+                        //
                         final LocalDate startOrEndDate;
                         if (billingMode == BillingMode.IN_ADVANCE) {
-                            startOrEndDate = invoiceItem.getStartDate();
-                        } else {
-                            startOrEndDate = invoiceItem.getStartDate().plus(billingPeriod.getPeriod());
+                            return invoiceItem.getStartDate().compareTo(cutoffDate) >= 0;
+                        } else /*  BillingMode.IN_ARREAR */ {
+                            if (invoiceItem.getEndDate().compareTo(cutoffDate) >= 0) {
+                                return true;
+                            }
+
+                            for (final Invoice inv : invoices) {
+                                final InvoiceItem existingItem = Iterables.tryFind(inv.getInvoiceItems(), new Predicate<InvoiceItem>() {
+                                    @Override
+                                    public boolean apply(final InvoiceItem item) {
+                                        return (item.getInvoiceItemType() == InvoiceItemType.RECURRING &&
+                                                item.getSubscriptionId().equals(invoiceItem.getSubscriptionId()) &&
+                                                item.getStartDate().compareTo(invoiceItem.getStartDate()) == 0);
+                                    }
+                                }).orNull();
+                                if (existingItem != null) {
+                                    if (existingItem.getEndDate().compareTo(invoiceItem.getEndDate()) != 0 ||
+                                        !existingItem.getPlanName().equals(invoiceItem.getPlanName()) ||
```

Comment:
Shouldn't we check the phase too?

Thinking out loud: is there any overlap with the `matches` logic? https://github.com/killbill/killbill/blob/00d9ace820bc7ddf6961a16a2727abe17b198bf9/invoice/src/main/java/org/killbill/billing/invoice/model/InvoiceItemCatalogBase.java#L125

Reply:
Added the check on phaseName as discussed in https://github.com/killbill/killbill/pull/1429/commits/5b280fc088678ff8acdf06647b29e1cabbd8deb1

Agree the logic is different than `InvoiceItemCatalogBase#matches` - I know for sure I need the end date in my case as I have no tree to create the boundaries, but *It seems like* checking on other changes for that same item (i.e same start date) regarding plan make sense, but this path is untested 🤷‍♂️